File: apps/external-apis/aldi-api/src/infrastructure/db/InMemoryProductRepository.ts
import { IProductRepository } from "../../domain/ports/IProductRepository";
import { Product } from "../../domain/entities/Product";
import { SearchProductsQuery } from "../../application/usecases/SearchProducts/SearchProductsQuerySchema";

export class InMemoryProductRepository implements IProductRepository {
    public products: Product[] = [];

    async search(query: SearchProductsQuery): Promise<{ items: Product[]; total: number; } | null> {
        const { query: searchQuery, page = 1, limit = 20 } = query;
        const normalizedQuery = searchQuery.toLowerCase();

        const filtered = this.products.filter(p => 
            p.title.toLowerCase().includes(normalizedQuery) ||
            p.description.toLowerCase().includes(normalizedQuery)
        );

        const total = filtered.length;
        const start = (page - 1) * limit;
        const end = start + limit;
        const items = filtered.slice(start, end);

        return { items, total };
    }

    // Helper for tests
    add(product: Product) {
        this.products.push(product);
    }
}

File: apps/external-apis/lidl-api/src/Lidl.Infrastructure/Repositories/InMemory/InMemoryProductRepository.cs
using Lidl.Application.Ports;
using Lidl.Application.UseCases.SearchProducts;
using Lidl.Domain.Entities;

namespace Lidl.Infrastructure.Repositories.InMemory;

public sealed class InMemoryProductRepository : IProductRepository
{
    private readonly List<Product> _items = [];

    public InMemoryProductRepository(IEnumerable<Product> items)
    {
        _items = items.ToList();
    }

    public Task<(IReadOnlyList<Product> Items, long Total)> SearchAsync(SearchProductsInput input, CancellationToken ct = default)
    {
        var q = _items.AsEnumerable();

        if (!string.IsNullOrWhiteSpace(input.Query))
            q = q.Where(p => p.Name.Contains(input.Query, StringComparison.OrdinalIgnoreCase));

        var total = q.LongCount();
        var items = q.Skip((input.Page - 1) * input.Limit).Take(input.Limit).ToList();

        return Task.FromResult(((IReadOnlyList<Product>)items, total));
    }
}